docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
```

Comment:
If all variables inside a role are prefixed with `r_` it will be hard to distinguish between variables that are meant to be local only and those that are meant to be read outside of a role. Again, we can make a list of all role variables that a role "produces" and document each one properly. However, nothing will stop a user from consuming the remaining ones. Of course, we can state" if you use any of the role variables that we have not declared as `output` variables, it is at your own risk chief". The `l_` prefix is here for that very reason. Of course, we don't have to use the `l_` prefix and can use `__some_variable_not_meant_to_be_used_outside_of_a_role` style instead. I would like to distinguish the concept of variables meant to be used within/outside a role on a syntactic level.